ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientRegistry.hs

```diff
@@ -164,24 +166,52 @@ bracketFetchClient (FetchClientRegistry ctxVar
                -> (ThreadId m, StrictTMVar m ())
                -> m ()
     unregister ksVar FetchClientContext { fetchClientCtxStateVars = stateVars }
-               (tid, doneVar)  = uninterruptibleMask_ $ do
-      -- Signal we are shutting down
-      atomically $
-        writeTVar (fetchClientStatusVar stateVars) PeerFetchStatusShutdown
-      -- Kill the sync client if it is still running
-      throwTo tid AsyncCancelled
-      -- Wait for the sync client to terminate and finally unregister ourselves
-      atomically $ do
-        -- Signal to keepAlive that we're going away
-        putTMVar ksVar ()
-        modifyTVar keepRegistry $ \m ->
-          assert (peer `Map.member` m) $
-          Map.delete peer m
+               (tid, doneVar)  = do
+      dead <- uninterruptibleMask_  $ do
+        -- Signal we are shutting down
+        dieFast <- atomically $ do
+          writeTVar (fetchClientStatusVar stateVars) PeerFetchStatusShutdown
+
+          dr <- readTVar dyingRegistry
+          return $ Set.member peer dr
+
+        -- If keepAlive is dying we don't need to let chainsync exit cleanly
```

Comment:
Why that's the case?

Reply:
Just as blockfetch kills chainsync, keepalive will kill blockfetch. If keepalive is dying this can't be a hot demotion.